src/EntityComponentManager.cc

```diff
@@ -894,24 +925,94 @@ void EntityComponentManager::ChangedState(
   // TODO(anyone) New / removed / changed components
 }
 
+void EntityComponentManager::CalculateComponentThreadLoad() const
+{
+  // If the entity component vector is dirty, we need to recalculate the
+  // threads and each threads work load
+  if (!this->dataPtr->entityComponentsDirty)
+    return;
+
+  this->dataPtr->entityComponentsDirty = false;
+  this->dataPtr->entityComponentIterators.clear();
+  auto startIt = this->dataPtr->entityComponents.begin();
+  int numComponents = this->dataPtr->entityComponents.size();
+
+  // Each thread runs `this->dataPtr->componentsPerThread` or more
+  // entity components, an arbitrary number
+  int numThreads = numComponents / this->dataPtr->componentsPerThread + 1;
+  int maxThreads = std::thread::hardware_concurrency();
+
+  // Set the number of threads to spawn to the min of the calculated thread
+  // count or max threads that the hardware supports
+  this->dataPtr->numComponentThreads = std::min(numThreads, maxThreads);
```

Comment:
I see what you're saying.  It's something I considered for awhile, but given the workload that the threads will need to process will change (if entity components are added or removed), we have to, at minimum, be able to maintain some record of each thread's allocated work, and that record will need to be updated every time the entity components are changed.  I think this is the fastest way of doing this that I can think of, the `std::next` call below is probably the costliest operation of the function given it is doing an `i++` iteration behind the scenes iterating through all the components in the map.

We could possibly have a static number of threads.  My main concern is the behavior when the entity component number is low.  As an example, a scenario when there are 5 entity components and 10 threads.  Would one thread handle all 5?  Would 5 threads each take one entity component and the other 5 are left doing nothing?  Additionally, since the main thread in `State()` would need to wait for all of the threads to finish doing their would, we would need to have some mechanism for the main thread to signal to the other threads which are waiting on that condition, and each of those threads would then need to have a mechanism to signal back to the main thread that they are done.  

This could be done, I believe, with `std::shared_lock`, `std::shared_mutex`, and `std::condition_variable`, having the main thread execute a `cv.notify_all()`.  Although I'm wondering how the main thread would be sure that all of the worker threads are done working.  I suppose we could have some atomic variable/condition that the main thread spins on until it has reached the number of threads.

So I guess it's just what we desire.  This initial approach in this PR was more or less the most straightforward way I could get the desired behavior in.  The benefits of the above described approach is only that it would kill the overhead of spawning the threads every iteration.  We would still need to have some sort of `dirty` boolean that notifies when we need to refresh each thread's assigned iterators.

Reply:
Thanks for the detailed info, John. I can see how using a fixed number of threads could be complicated.

> Would 5 threads each take one entity component and the other 5 are left doing nothing?

Yeah my intuition would be to leave the unnecessary threads idling until they're needed. I think this would be a minority of use cases, right?

> I suppose we could have some atomic variable/condition that the main thread spins on until it has reached the number of threads.

I think that the [Barrier](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/Barrier.hh) class could be helpful here. That's being used to synchronize the post-update threads:

https://github.com/ignitionrobotics/ign-gazebo/blob/a830b443989877b1e83170390f483b6c02952b25/src/SimulationRunner.cc#L408-L439

Reply:
Sounds good, I updated the `componentsPerThread` to 1 2e253210ed78695db9da45ffb8e025d9985c6d75 .  This will allocate one entity component per thread up to the max number of threads, then it will begin evenly allocating components to each thread.